### PR TITLE
Draft: Extend flattened storage

### PR DESCRIPTION
**Pull request by Leimeroth**
Function to extend a flattened storage with another FlattenedStorage.

@pmrv Would be greate to get some feedback on this soon, because I would like to use it to make the Atomicrex job compatible with PotentialFit

### CODE REVIEW

File: pyiron_base/generic/flattenedstorage.py [hunk 1]
```diff
@@ -18,7 +18,9 @@
 __date__ = "Jul 16, 2020"
 
 
+import chunk
 import copy
+from multiprocessing.sharedctypes import Value
```
Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py]:
```suggestion
import copy
```
I suppose this is left over from something else.

File: pyiron_base/generic/flattenedstorage.py [hunk 2]
```diff
@@ -796,6 +798,60 @@ def add_chunk(self, chunk_length, identifier=None, **arrays):
         self.current_element_index = i
         # return last_chunk_index, last_element_index
 
+    def extend_storage(self, other: FlattenedStorage):
```
Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py]:
```suggestion
    def extend(self, other: FlattenedStorage):
```

File: pyiron_base/generic/flattenedstorage.py [hunk 3]
```diff
@@ -796,6 +798,60 @@ def add_chunk(self, chunk_length, identifier=None, **arrays):
         self.current_element_index = i
         # return last_chunk_index, last_element_index
 
+    def extend_storage(self, other: "FlattenedStorage"):
```
Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py:801]:
```suggestion
    def extend(self, other: "FlattenedStorage"):
```

File: pyiron_base/generic/flattenedstorage.py [hunk 4]
```diff
@@ -796,6 +798,60 @@ def add_chunk(self, chunk_length, identifier=None, **arrays):
...
+        combined_num_elements = self.num_elements + other.num_elements
+        if combined_num_chunks > self._num_chunks_alloc:
+            self._resize_chunks(combined_num_chunks)
+        if combined_num_elements > self._num_elements_alloc:
+            self._resize_elements(combined_num_elements)
+
+        for k, a in other._per_chunk_arrays.items():
+            # add start_index of last chunk to start_index of other for correct mapping
+            if k == "start_index":
+                a += self._per_chunk_arrays[k][self.num_chunks]
```
Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py]: I think this modifies the `other` storage in place.  Double check that in tests.

File: pyiron_base/generic/flattenedstorage.py [hunk 5]
```diff
@@ -895,3 +951,20 @@ def read_array(name, hdf):
 
         if version >= "0.3.0":
             self._fill_values = hdf["_fill_values"]
+
+
+def _get_dtype_and_fill(storage: FlattenedStorage, name: str):
```
Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py]:
```suggestion
def _get_dtype_and_fill(storage: FlattenedStorage, name: str) -> (np.generic, Any):
```

File: pyiron_base/generic/flattenedstorage.py [hunk 6]
```diff
@@ -796,6 +798,60 @@ def add_chunk(self, chunk_length, identifier=None, **arrays):
...
+                self.add_array(name=k, dtype=dtype, fill=fill, per="element")
+            self._per_element_arrays[k][
+                self.num_elements : combined_num_elements
+            ] = other._per_element_arrays[k][0 : other.num_elements]
+        self.num_elements = combined_num_elements
+        self.num_chunks = combined_num_chunks
+        self.current_chunk_index = self.num_chunks
+        self.current_element_index = self.num_elements
+
+    def _check_compatible_fill_values(self, other: "FlattenedStorage"):
```
Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py:844]: Can be a proper function outside the class.

File: pyiron_base/generic/flattenedstorage.py [hunk 7]
```diff
@@ -796,6 +798,60 @@ def add_chunk(self, chunk_length, identifier=None, **arrays):
...
+        f.e. after extending to the storage.
+
+        Args:
+            other (FlattenedStorage): Another FlattenedStorage instance
+
+        Raises:
+            ValueError: Raises when the storages have different fill values for a key
+        """
+        for k in self._fill_values.keys():
+            if k in other._fill_values.keys():
```
Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py]:
```suggestion
        for k in self._fill_values:
            if k in other._fill_values:
```

Comment (pmrv) [inline pyiron_base/generic/flattenedstorage.py]:
```suggestion
        for k in set(self._fill_values).intersection(other._fill_values):
```

Comment (pmrv) [reply pyiron_base/generic/flattenedstorage.py]: The second one saves a layer of indentation, which I find nice usually.

**Review by pmrv — APPROVED** [review]
Some nits, but looks good to me otherwise.  Please add tests though before converting from draft.

File: pyiron_base/generic/flattenedstorage.py (hunk 1 above)
Comment (Leimeroth) [reply pyiron_base/generic/flattenedstorage.py]: Maybe some were autoimports from some pseudo-smart code extension

**Pull request merged**